packages/contracts/contracts/Unirep.sol

```diff
@@ -239,6 +239,38 @@ contract Unirep is IUnirep, zkSNARKHelper, VerifySignature {
         );
     }
 
+    function submitGSTAttestation(
```

Comment:
How about we also overload `submitAttestation`
They are too similar in `_submitAttestation`

or we let it be an optional field in `submitAttestation`
if `gstRoot=0` then check `fromProofIndex` and `toProofIndex`, else the synchronizer checks `gstRoot`

the AttestationSubmitted event can be
```solidity
event AttestationSubmitted(
    uint256 indexed epoch,
    uint256 indexed epochKey,
    address indexed attester,
    AttestationEvent attestationEvent,
    Attestation attestation,
    uint256 toProofIndex,  // optional
    uint256 fromProofIndex, // optional
    uint256 gstRoot // optional
);
```

Or I would prefer we even remove the `fromProofIndex` and `toProofIndex` and `gstRoot` to make the unirep the most lightweight and attesters should be responsible for all reputation they give

Moreover, the hashchain can happen on chain. It is what the previous version unirep did.
(but this is open to discussion)

Reply:
> if gstRoot=0 then check fromProofIndex and toProofIndex, else the synchronizer checks gstRoot

I think we can use a modifier to do the common checks for attestations. I really don't like taking multiple variables and doing lots of if logic to do certain things. Functions should do 1 thing.

> the AttestationSubmitted event can be

I don't think combining the events is a good thing. It makes the attestation processing function very complex off chain. It also requires each event to have fields that are unused.

Also bear in mind `AttestationEvent attestationEvent` is removed in #87. We don't need it because the proof events can identify the type of proof.

> Or I would prefer we even remove the fromProofIndex and toProofIndex and gstRoot to make the unirep the most lightweight and attesters should be responsible for all reputation they give

I don't understand how this would be possible. Would attesters emit proof indexes and gst roots?